### DISCUSSION TITLE
Copy watchstate from UserA to UserB in the same Jellyfin server

### DISCUSSION THREAD

**Discussion by fa1rid** [arabcoders/watchstate#562] [General] [answered]
Is that possible? If yes, how please?

This is me first time using the tool. I tried to do an import but it doesn't seem to work, and I don't know how to debug the issue:

```
~ console state:import -v -s UserA
[2024-11-04T00:25:41+00:00] NOTICE: Using WatchState Version - '20241012-16e9e36'.
[2024-11-04T00:25:41+00:00] NOTICE: SYSTEM: Preloading 'MemoryMapper' data.
[2024-11-04T00:25:41+00:00] NOTICE: SYSTEM: Preloading 'MemoryMapper' data completed in '0s'.
[2024-11-04T00:25:41+00:00] NOTICE: SYSTEM: Importing 'UserA' metadata & play state changes.
[2024-11-04T00:25:41+00:00] WARNING: No requests for 'Jellyfin: UserA' libraries were queued.
[2024-11-04T00:25:41+00:00] WARNING: SYSTEM: Not updating 'UserA' import last sync date. There was errors recorded during the operation.
[2024-11-04T00:25:41+00:00] NOTICE: SYSTEM: Waiting on '0' requests.
[2024-11-04T00:25:41+00:00] NOTICE: SYSTEM: Completed waiting on '0' requests in '0's. Parsed '0.00B' of data.
[2024-11-04T00:25:41+00:00] NOTICE: MAPPER: No changes detected.
┌─────────┬───────┬─────────┬────────┐
│ Type    │ Added │ Updated │ Failed │
├─────────┼───────┼─────────┼────────┤
│ Movie   │ 0     │ 0       │ 0      │
├─────────┼───────┼─────────┼────────┤
│ Episode │ 0     │ 0       │ 0      │
└─────────┴───────┴─────────┴────────┘
```

**Update**

I learned to add the `--debug --logfile /config/export.txt` option, so now I can see that my libraries are being marked as type 'unknown' in the log (attached as `import.log`). Can you explain why this is happening and how to correct it?
[import.log](https://github.com/user-attachments/files/17612992/import.log)

**Update 2**

I'm considering updating the database directly using SQLite. I found a table named `UserDatas` which holds the watch states for all users. For each item, I noticed there are three entries. 

Example for a single movie, (keys) column:
- 38700
- tt1502397
- a1523bf5-a01a-1f3b-1b1d-15d1ba053326

The first seems to be the TMDB ID, the second the IMDB ID, but I’m not sure what the third is.

In the `TypedBaseItems` table, I found that the `UserDataKey` column for this movie shows `38700` (the TMDB ID), which seems to link it to the `UserDatas` table. I'm still unclear on why each item has three records.

My plan is to export all records for the user I need from the `UserDatas` table, replace the user ID with the new one, and then re-import the data. Would this approach work?

**Comment by arabcoders** [comment, marked as answer]
Hi, from the debug log your libraries don't have `CollectionType` i am not sure why. we only support movies and series collection type. 

For your 2nd question sadly i am not sure it will work directly by modifying the data directly as the ids differ between users afaik. 

If you fix the library content type everything should work.

**Comment by fa1rid** [comment]
It seems there is a bug in my database for some reason. The libraries were created originally correctly as Movies and Shows, but seems Jellyfin updates messed up the database or something.

Though, they are showing correctly with their type in Jellyfin dashboard:

![image](https://github.com/user-attachments/assets/ca9c2e9e-11d6-4556-ab80-09135b553914)
![image](https://github.com/user-attachments/assets/df6b76c7-757f-48b0-86c3-677d50ba8d04)

So that's really strange, right?


I can actually just run this query without the need to export/import:
Copy watch state from userId 1 to 4

```
INSERT INTO UserDatas (key, userId, rating, played, playCount, isFavorite, playbackPositionTicks, lastPlayedDate, AudioStreamIndex, SubtitleStreamIndex)
SELECT key, 4 AS userId, rating, played, playCount, isFavorite, playbackPositionTicks, lastPlayedDate, AudioStreamIndex, SubtitleStreamIndex
FROM UserDatas
WHERE userId = 1;
```

What do you mean ids differ between users? I'm 99% sure this will work. I will try and let you know.

Anyway, is it possible to copy watchstate from UserA to UserB in the same Jellyfin server using your tool?

**Reply by arabcoders** [reply]
It seems your db has problems i am not sure why you probably need to go and edit the library and make sure the content type is set correctly. 

And about the manual queries i don't recommend manually altering however, backup the db and try i guess :-) 

And yes you can add as many users as you like. however it should be one way otherwise it will alter your state. 

i.e, import from userA, and userB/C should be export target only no import only metadata.

**Comment by fa1rid** [comment]
There is no option to edit the library, I guess you cannot change the type after it's created. But as you saw it's showing correctly in the dashboard.

The query worked! I can see all the watch states now in the new user.

I guess I need to create an issue with Jellyfin to see why I have the issue in my setup.

**Comment by fa1rid** [comment]
Check this: https://github.com/jellyfin/jellyfin/issues/12958

**Reply by arabcoders** [reply]
> Check this: https://github.com/jellyfin/jellyfin/issues/12958

As i said it's the `CollectionType` not `MediaType` the library is a collection not a media. 

You can check the source related to the check in https://github.com/arabcoders/watchstate/blob/16e9e36c12d158ef75bc1ee8ac5db1373d9da2e6/src/Backends/Jellyfin/Action/GetLibrariesList.php#L115-L154

**Comment by fa1rid** [comment]
OMG, I misread that. Thanks for clarification. I will follow up with the Jellyfin issue with the correction. Will update you once I know what is the issue.

**Reply by arabcoders** [reply]
No worries. Probably have something to do with a plugin you used before.
